apps/glusterfs/volume_entry_allocate.go

```diff
@@ -63,6 +64,165 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
 	}
 }
 
+func (v *VolumeEntry) replaceBrickInVolume(db *bolt.DB, executor executors.Executor,
```

Comment:
This function is _the_ most important part of the fix.  It needs to make sure that it picks a new brick from the ring that is on a _different_ zone than the ones already used and/or in the same zone if there is only one zone.  The algorithm used here should be discussed first.